milli/src/update/index_documents/transform.rs

```diff
@@ -151,204 +126,99 @@ impl Transform<'_, '_> {
-                });
-            }
-
-            // We extract the user id if we know where it is or generate an UUID V4 otherwise.
-            let external_id = match primary_key_pos {
-                Some(pos) => {
-                    let external_id = &record[pos];
-                    // We validate the document id [a-zA-Z0-9\-_].
-                    match validate_document_id(&external_id) {
-                        Some(valid) => valid,
-                        None => {
-                            return Err(UserError::InvalidDocumentId {
-                                document_id: Value::from(external_id),
+                            if let Some((k, v)) = key.zip(value) {
+                                json.insert(k, v);
                             }
-                            .into())
                         }
+
+                        return Err(UserError::MissingDocumentId { document: json }.into());
                     }
+
+                    let uuid = uuid::Uuid::new_v4().to_hyphenated().encode_lower(&mut uuid_buffer);
+                    serde_json::to_writer(&mut external_id_buffer, &uuid).unwrap();
+                    field_buffer.push((primary_key_id, &external_id_buffer));
+                    Cow::Borrowed(&*uuid)
                 }
-                None => uuid::Uuid::new_v4().to_hyphenated().encode_lower(&mut uuid_buffer),
             };
 
-            // When the primary_key_field_id is found in the fields ids list
-            // we return the generated document id instead of the record field.
-            let iter = fields_ids.iter().map(|(fi, i)| {
-                let field = if *fi == primary_key_id { external_id } else { &record[*i] };
-                (fi, field)
-            });
+            // Insertion in a obkv need to be done with keys ordered. For now they are ordered
+            // according to the document addition key order, so we sort it according to the
+            // fieldids map keys order.
+            field_buffer.sort_unstable_by(|(f1, _), (f2, _)| f1.cmp(&f2));
 
-            // We retrieve the field id based on the fields ids map fields ids order.
-            for (field_id, field) in iter {
-                // We serialize the attribute values as JSON strings.
-                json_buffer.clear();
-                serde_json::to_writer(&mut json_buffer, &field)
-                    .map_err(InternalError::SerdeJson)?;
-                writer.insert(*field_id, &json_buffer)?;
+            // The last step is to build the new new obkv document, and insert it in the sorter.
+            let mut writer = obkv::KvWriter::new(&mut obkv_buffer);
+            for (k, v) in field_buffer.iter() {
+                writer.insert(*k, v)?;
             }
 
             // We use the extracted/generated user id as the key for this document.
-            sorter.insert(external_id, &obkv_buffer)?;
+            sorter.insert(&external_id.as_ref().as_bytes(), &obkv_buffer)?;
```

Comment:
```suggestion
            sorter.insert(&external_id, &obkv_buffer)?;
```

Not sure why you need to do all of this, but maybe just doing this is OK as the insert method accepts an `AsRef<[u8]>`.

Reply:
Cow<&str> doesn't implement AsRef<u8> unfortunately

Reply:
`Cow<str>` does implement `AsRef<str>` which it self implements `Cow<[u8]>`, but is [it better](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=bcd13d7ac7e6ddf4d1c1a5c46a515ba1)?